internal/server/server_test.go

```diff
@@ -1,47 +1,96 @@
 package server
 
 import (
+	"log"
 	"net/http"
 	"net/http/httptest"
 	"os"
 	"testing"
 
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/gorilla/sessions"
 	"github.com/stretchr/testify/assert"
 )
 
-var testHandler http.Handler
+var server *Server
 
 func TestMain(m *testing.M) {
 	var err error
-	testHandler, err = New()
+	server, err = New(
+		WithMockOIDCNewProvider(MockOIDCNewProvider),
+	)
 	if err != nil {
+		log.Printf("server.New failed, error: %v", err)
 		os.Exit(1)
 	}
 
+	server.oauth2Config = &MockOauth2Config{}
+	server.verifier = oidc.NewVerifier("", &MockKeySet{}, &oidc.Config{ClientID: server.kubectlClientID})
+
 	os.Exit(m.Run())
 }
 
 func TestHandleRoot(t *testing.T) {
 	t.Parallel()
 
 	w := httptest.NewRecorder()
-	testHandler.ServeHTTP(w, httptest.NewRequest("GET", "/", nil))
+	server.ServeHTTP(w, httptest.NewRequest("GET", "/", nil))
 	assert.Equal(t, http.StatusOK, w.Code)
 }
 
 func TestHandleLoginGet(t *testing.T) {
 	t.Parallel()
 
 	w := httptest.NewRecorder()
-	testHandler.ServeHTTP(w, httptest.NewRequest("GET", "/login", nil))
+	server.ServeHTTP(w, httptest.NewRequest("GET", "/login", nil))
 	assert.Equal(t, http.StatusMethodNotAllowed, w.Code)
 }
 
 func TestHandleLoginPost(t *testing.T) {
 	t.Parallel()
 
 	w := httptest.NewRecorder()
-	testHandler.ServeHTTP(w, httptest.NewRequest("POST", "/login", nil))
-	assert.Equal(t, http.StatusOK, w.Code)
+	server.ServeHTTP(w, httptest.NewRequest("POST", "/login", nil))
+	assert.Equal(t, http.StatusSeeOther, w.Code)
 	assert.NotEmpty(t, w.Result().Cookies()[0])
 }
+
+func TestHandleCallbackGet(t *testing.T) {
+	t.Parallel()
+
+	var (
+		w       *httptest.ResponseRecorder
+		req     *http.Request
+		session *sessions.Session
+	)
+
+	const validCallbackUrl = "/callback?code=testcode"
+
+	// Success path
+	w = httptest.NewRecorder()
+	req = httptest.NewRequest("GET", validCallbackUrl, nil)
+	session = server.getSession(req)
+	session.Values["nonce"] = ""
```

Comment:
This test seems like it should be able to set a valid nonce.

I am seeing a few errors in the logs related to nonce that I have not been able to trace back yet.

```
time="2021-09-28T12:01:20-04:00" level=error msg="id token nonce does not match session nonce" error="error invalid nonce"
```

Reply:
Fixed